File: services/media_pipeline/model_profiles.py
import os

from bootstrap.path_layout import get_project_root


BACKEND_DIR = os.path.dirname(os.path.abspath(__file__))
APP_ROOT = get_project_root(BACKEND_DIR)
MODELS_ROOT = os.path.join(APP_ROOT, "models")
USER_HOME = os.path.expanduser("~")


DEFAULT_ASR_MODEL_PROFILES = {
    "faster-whisper": "quality",
    "funasr": "standard",
    "qwen": "standard",
    "vibevoice-asr": "standard",
    "jianying": "default",
    "bcut": "default",
}

DEFAULT_TTS_MODEL_PROFILES = {
    "indextts": "standard",
    "qwen": "quality",
}


ASR_MODEL_PROFILES = {
    "funasr": {
        "standard": {
            "label": "SenseVoiceSmall + fsmn-vad",
            "model_name": "SenseVoiceSmall",
            "modelscope_model_id": "iic/SenseVoiceSmall",
            "hf_model_id": "FunAudioLLM/SenseVoiceSmall",
            "vad_model_name": "fsmn-vad",
            "modelscope_vad_id": "iic/speech_fsmn_vad_zh-cn-16k-common-pytorch",
            "use_itn": True,
            "candidates": [
                os.path.join(MODELS_ROOT, "FunASR-SenseVoiceSmall"),
                os.path.join(MODELS_ROOT, "SenseVoiceSmall"),
                os.path.join(USER_HOME, ".cache", "modelscope", "hub", "models", "iic", "SenseVoiceSmall"),
            ],
            "vad_candidates": [
                os.path.join(MODELS_ROOT, "FunASR-fsmn-vad"),
                os.path.join(USER_HOME, ".cache", "modelscope", "hub", "models", "iic", "speech_fsmn_vad_zh-cn-16k-common-pytorch"),
            ],
            "languages": ["auto", "zh", "yue", "en", "ja", "ko", "nospeech"],
        },
        "zh": {
            "label": "paraformer-zh + fsmn-vad + ct-punc",
            "model_name": "paraformer-zh",
            "modelscope_model_id": "iic/speech_paraformer-large-vad-punc_asr_nat-zh-cn-16k-common-vocab8404-pytorch",
            "vad_model_name": "fsmn-vad",
            "modelscope_vad_id": "iic/speech_fsmn_vad_zh-cn-16k-common-pytorch",
            "punc_model_name": "ct-punc",
            "modelscope_punc_id": "iic/punc_ct-transformer_zh-cn-common-vocab272727-pytorch",
            "languages": ["auto", "zh"],
            "use_itn": True,
            "candidates": [
                os.path.join(MODELS_ROOT, "FunASR-paraformer-zh"),
            ],
            "vad_candidates": [
                os.path.join(MODELS_ROOT, "FunASR-fsmn-vad"),
                os.path.join(USER_HOME, ".cache", "modelscope", "hub", "models", "iic", "speech_fsmn_vad_zh-cn-16k-common-pytorch"),
            ],
            "punc_candidates": [
                os.path.join(MODELS_ROOT, "FunASR-ct-punc"),
            ],
        },
    },
    "faster-whisper": {
        "balanced": {
            "label": "large-v3-turbo",
            "model_name": "large-v3-turbo",
            "candidates": [
                os.path.join(MODELS_ROOT, "faster-whisper-large-v3-turbo-ct2"),
                os.path.join(MODELS_ROOT, "faster-whisper-large-v3-ct2"),
            ],
        },
        "quality": {
            "label": "large-v3",
            "model_name": "large-v3",
            "candidates": [
                os.path.join(MODELS_ROOT, "faster-whisper-large-v3-ct2"),
                os.path.join(MODELS_ROOT, "faster-whisper-large-v3-turbo-ct2"),
            ],
        },
    },
    "qwen": {
        "standard": {
            "label": "Qwen3-ASR-1.7B",
            "model_name": "Qwen3-ASR-1.7B",
            "candidates": [
                os.path.join(MODELS_ROOT, "Qwen3-ASR-1.7B"),
            ],
        },
        "fast": {
            "label": "Qwen3-ASR-0.6B",
            "model_name": "Qwen3-ASR-0.6B",
            "candidates": [
                os.path.join(MODELS_ROOT, "Qwen3-ASR-0.6B"),
            ],
        },
    },
    "vibevoice-asr": {
        "standard": {
            "label": "VibeVoice-ASR-HF",
            "model_name": "VibeVoice-ASR-HF",
            "hf_model_id": "microsoft/VibeVoice-ASR-HF",
            "candidates": [
                os.path.join(MODELS_ROOT, "VibeVoice-ASR-HF"),
            ],
        },
    },
}


TTS_MODEL_PROFILES = {
    "indextts": {
        "standard": {
            "label": "Index-TTS Standard",
            "model_dir_candidates": [
                os.path.join(MODELS_ROOT, "index-tts"),
            ],
        },
    },
    "qwen": {
        "quality": {
            "label": "Qwen3-TTS 1.7B",
            "qwen_model_size": "1.7B",
        },
        "fast": {
            "label": "Qwen3-TTS 0.6B",
            "qwen_model_size": "0.6B",
        },
    },
}


def _normalize_key(value):
    return str(value or "").strip().lower().replace("_", "-")


def normalize_asr_model_profile(service, requested_profile=None):
    service_key = _normalize_key(service)
    profiles = ASR_MODEL_PROFILES.get(service_key)
    if not profiles:
        return DEFAULT_ASR_MODEL_PROFILES.get(service_key, "default")

    requested_key = _normalize_key(requested_profile)
    if requested_key in profiles:
        return requested_key
    return DEFAULT_ASR_MODEL_PROFILES.get(service_key, next(iter(profiles.keys())))


def normalize_tts_model_profile(service, requested_profile=None):
    service_key = _normalize_key(service)
    profiles = TTS_MODEL_PROFILES.get(service_key)
    if not profiles:
        return DEFAULT_TTS_MODEL_PROFILES.get(service_key, "default")

    requested_key = _normalize_key(requested_profile)
    if requested_key in profiles:
        return requested_key
    return DEFAULT_TTS_MODEL_PROFILES.get(service_key, next(iter(profiles.keys())))


def get_asr_profile(service, requested_profile=None):
    service_key = _normalize_key(service)
    profiles = ASR_MODEL_PROFILES.get(service_key, {})
    profile_key = normalize_asr_model_profile(service_key, requested_profile)
    return profile_key, profiles.get(profile_key, {})


def get_tts_profile(service, requested_profile=None):
    service_key = _normalize_key(service)
    profiles = TTS_MODEL_PROFILES.get(service_key, {})
    profile_key = normalize_tts_model_profile(service_key, requested_profile)
    return profile_key, profiles.get(profile_key, {})


def resolve_existing_path(candidates):
    for candidate in candidates or []:
        if candidate and os.path.exists(candidate):
            return candidate
    return None

File: apps/desktop/ui/src/utils/modelProfiles.ts
import type { AsrService } from './asrService';

export type TtsService = 'indextts' | 'qwen' | 'gptsovits';

export interface ModelProfileOption {
    id: string;
    label: string;
    description: string;
}

export const ASR_MODEL_PROFILES: Record<AsrService, ModelProfileOption[]> = {
    bcut: [
        { id: 'default', label: '云端默认', description: '由服务端自动选择，无需本地模型。' }
    ],
    jianying: [
        { id: 'default', label: '云端默认', description: '由服务端自动选择，无需本地模型。' }
    ],
    funasr: [
        { id: 'standard', label: 'Standard / SenseVoiceSmall', description: '多语言档位，适合中英混合与英文素材。' },
        { id: 'zh', label: 'Chinese / paraformer-zh', description: '中文优先档位，使用 paraformer-zh、fsmn-vad 和 ct-punc。' }
    ],
    'faster-whisper': [
        { id: 'quality', label: 'Quality / large-v3', description: '高质量档位，适用于正式字幕生产。' },
        { id: 'balanced', label: 'Balanced / large-v3-turbo', description: '均衡档位，兼顾处理速度与识别质量。' }
    ],
    qwen: [
        { id: 'standard', label: 'Standard / 1.7B', description: '标准档位，面向多语种正式识别任务。' },
        { id: 'fast', label: 'Fast / 0.6B', description: '轻量档位，适合快速处理与受限设备。' }
    ],
    'vibevoice-asr': [
        { id: 'standard', label: 'Standard / HF', description: '标准档位，适用于长音频与多说话人内容。' }
    ]
};

export const TTS_MODEL_PROFILES: Record<TtsService, ModelProfileOption[]> = {
    indextts: [
        { id: 'standard', label: 'Standard', description: '标准档位，适用于常规语音合成任务。' }
    ],
    gptsovits: [
        { id: 'balanced', label: 'Balanced / 平衡', description: '默认生产档位，兼顾速度、稳定性与完整朗读。' },
        { id: 'fast', label: 'Fast / 极速', description: '优先吞吐与响应速度，自动保持在硬件安全范围内。' },
        { id: 'quality', label: 'Quality / 高保真', description: '优先文本完整度与发音稳定性，适合正式成片。' }
    ],
    qwen: [
        { id: 'quality', label: 'Quality / 1.7B', description: '高质量档位，适用于正式配音生产。' },
        { id: 'fast', label: 'Fast / 0.6B', description: '轻量档位，适合快速试听与资源受限设备。' }
    ]
};

const ASR_PROFILE_STORAGE_KEYS: Record<AsrService, string> = {
    bcut: 'asr_model_profile_bcut',
    jianying: 'asr_model_profile_jianying',
    funasr: 'asr_model_profile_funasr',
    'faster-whisper': 'asr_model_profile_faster_whisper',
    qwen: 'asr_model_profile_qwen',
    'vibevoice-asr': 'asr_model_profile_vibevoice_asr'
};

const TTS_PROFILE_STORAGE_KEYS: Record<TtsService, string> = {
    indextts: 'tts_model_profile_indextts',
    gptsovits: 'tts_model_profile_gptsovits',
    qwen: 'tts_model_profile_qwen'
};

export function getDefaultAsrModelProfile(service: AsrService): string {
    return ASR_MODEL_PROFILES[service][0]?.id || 'default';
}

export function getDefaultTtsModelProfile(service: TtsService): string {
    return TTS_MODEL_PROFILES[service][0]?.id || 'standard';
}

export function getStoredAsrModelProfile(service: AsrService): string {
    const saved = localStorage.getItem(ASR_PROFILE_STORAGE_KEYS[service]);
    const options = ASR_MODEL_PROFILES[service];
    if (saved && options.some(option => option.id === saved)) {
        return saved;
    }
    return getDefaultAsrModelProfile(service);
}

export function setStoredAsrModelProfile(service: AsrService, profileId: string) {
    localStorage.setItem(ASR_PROFILE_STORAGE_KEYS[service], profileId);
}

export function getStoredTtsModelProfile(service: TtsService): string {
    const saved = localStorage.getItem(TTS_PROFILE_STORAGE_KEYS[service]);
    const options = TTS_MODEL_PROFILES[service];
    if (saved && options.some(option => option.id === saved)) {
        return saved;
    }
    return getDefaultTtsModelProfile(service);
}

export function setStoredTtsModelProfile(service: TtsService, profileId: string) {
    localStorage.setItem(TTS_PROFILE_STORAGE_KEYS[service], profileId);
    if (service === 'qwen') {
        localStorage.setItem('qwen_tts_model', profileId === 'fast' ? '0.6B' : '1.7B');
    }
}
